crates/matrix-sdk-crypto/src/dehydrated_devices.rs

```diff
@@ -0,0 +1,532 @@
+    /// // Don't use a zero key for real.
+    /// let pickle_key = [0u8; 32];
+    ///
+    /// // Create the dehydrated device.
+    /// let device = machine.dehydrated_devices().create();
+    ///
+    /// // Create the request that should upload the device.
+    /// let request = device
+    ///     .keys_for_upload("Dehydrated device".to_owned(), &pickle_key)
+    ///     .await?;
+    ///
+    /// // Send the request out using your HTTP client.
+    /// // client.send(request).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[instrument(
+        skip_all, fields(
+            user_id = %self.account.user_id(),
+            device_id = %self.account.device_id(),
+            identity_keys = ?self.account.identity_keys,
+        )
+    )]
+    pub async fn keys_for_upload(
+        &self,
+        initial_device_display_name: String,
+        pickle_key: &[u8; 32],
+    ) -> Result<put_dehydrated_device::unstable::Request, DehydrationError> {
+        self.account.generate_fallback_key_helper().await;
+        let (device_keys, one_time_keys, fallback_keys) = self.account.keys_for_upload().await;
+
+        let mut device_keys = device_keys
+            .expect("We should always try to upload device keys for a dehydrated device.");
+
+        self.account
+            .store
+            .private_identity()
+            .lock()
+            .await
+            .sign_device_keys(&mut device_keys)
+            .await?;
+
+        trace!("Creating a upload request for a dehydrated device");
+
+        let pickle_key = expand_pickle_key(pickle_key, self.account.device_id());
+        let device_id = self.account.device_id().to_owned();
+        let device_data = self.account.dehydrate(&pickle_key).await;
+
+        Ok(
+            assign!(put_dehydrated_device::unstable::Request::new(device_id, initial_device_display_name, device_data, device_keys.to_raw()), {
+                one_time_keys, fallback_keys
+            }),
+        )
+    }
+}
+
+/// We're using the libolm compatible pickle format and its encryption scheme.
+///
+/// The libolm pickle encryption scheme uses HKDF to deterministically expand an
+/// input key material, usually 32 bytes, into a AES key, MAC key, and the
+/// initialization vector (IV).
+///
+/// This means that the same input key material will always end up producing the
+/// same AES key, and IV.
+///
+/// This encryption scheme is used in the Olm double ratchet and was designed to
+/// minimize the size of the ciphertext. As a tradeof, it requires a unique
+/// input key material for each plaintext that gets encrypted, otherwise IV
+/// reuse happens.
+///
+/// To combat the IV reuse, we're going to create a per-dehydrated-device unique
+/// pickle key by expanding the key itself with the device ID used as the salt.
+///
+/// TODO: Perhaps we should put this into vodozemac with a new pickle
+/// minimalistic pickle format using the [`matrix_pickle`] crate.
```

Comment:
Should this really be part of the docs? Maybe put it as a regular comment as the first thing in the function body?

Reply:
It's a method private to this module, not sure it makes any difference, but sure can move it.